src/algo/tensor_path.py

```diff
@@ -0,0 +1,178 @@
+from pygraphblas import Matrix
+from more_itertools import unique_everseen
+from itertools import islice
+
+from src.grammar.rsa import RecursiveAutomaton
+from src.graph.label_graph import LabelGraph
+
+class Paths:
+
+    def __init__(self):
+        self.paths = []
+        self.path_start = dict()
+        self.path_end = dict()
+
+    def add_one_edge(self, path):
+        self.paths.append([path])
+
+        self.path_end.update({len(self.paths) - 1: path[1]})
+
+        try:
+            self.path_start[path[0]].append(len(self.paths) - 1)
+        except KeyError:
+            self.path_start.update({path[0]: [len(self.paths) - 1]})
+
+    def union_paths(self, second):
+
+        if len(second.paths) == 0:
+            return
+
+        for i in range(len(second.paths)):
+            self.path_end.update({len(self.paths) + i: second.path_end[i]})
+        for start in second.path_start:
+            second.path_start.update({start: [i + len(self.paths) for i in second.path_start[start]]})
+
+            if start in self.path_start:
+                self.path_start[start].extend(second.path_start[start])
+            else:
+                self.path_start.update({start: second.path_start[start]})
+
+        self.paths.extend(second.paths)
+
+    def clean_paths(self):
+        self.paths.clear()
+
+    def doSet(self):
+        self.paths = list(unique_everseen(self.paths, key=tuple))
+
+    def product_paths(self, right):
+
+        if not self.paths:
+            for path_r in right.paths:
+                self.paths.append(path_r)
+
+            self.path_end = right.path_end
+            self.path_start = right.path_start
+            right.clean_paths()
+
+        if self.paths and right.paths:
+            for pos_l in self.path_end:
+                if self.path_end[pos_l] in right.path_start:
+                    for pos_r in right.path_start[self.path_end[pos_l]]:
+                        self.paths[pos_l].extend(right.paths[pos_r])
+                        self.path_end.update({pos_l: right.path_end[pos_r]})
+
+
+class TensorPaths:
```

Comment:
I think we should extract some common interface for algorithms, which can solve CFPQ problem, like [here](https://github.com/JetBrains-Research/CFPQ_PyAlgo/blob/dd0ecececd79caa3024205fd1d8244d0154899cb/src/algo/single_source/single_source.py#L47) and common interface for query representation (`CnfGrammar` and `RecursiveAutomaton` should implement the same interface). After that we can write polymorphic tests.